cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java

```diff
@@ -0,0 +1,252 @@
+
+      // need to get old property and check if it contains profile information
+      Map<String, String> oldProperties = preferencesDataset.getPreferences(entityId);
+      preferencesDataset.setPreferences(entityId, propertyMap);
+      NamespaceId namespaceId =
+        entityId instanceof NamespacedEntityId ? ((NamespacedEntityId) entityId).getNamespaceId() : NamespaceId.SYSTEM;
+
+      // validate the profile and publish the necessary metadata change if the profile exists in the property
+      Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(namespaceId, propertyMap);
+      // get the old profile information from the previous properties
+      Optional<ProfileId> oldProfile = SystemArguments.getProfileIdFromArgs(namespaceId, oldProperties);
+      if (profile.isPresent()) {
+        ProfileId profileId = profile.get();
+        if (profileDataset.getProfile(profileId).getStatus() == ProfileStatus.DISABLED) {
+          throw new ProfileConflictException(String.format("Profile %s in namespace %s is disabled. It cannot be " +
+                                                             "assigned to any programs or schedules",
+                                                           profileId.getProfile(), profileId.getNamespace()),
+                                             profileId);
+        }
+        adminEventPublisher.publishProfileAssignment(messagingContext, entityId, profileId);
+      }
+
+      // if new profiles do not have profile information but old profiles have, it is same as deletion of the profile
+      if (!profile.isPresent() && oldProfile.isPresent()) {
+        adminEventPublisher.publishProfileUnAssignment(messagingContext, entityId);
+      }
+    }, NotFoundException.class, ProfileConflictException.class);
+  }
+
+  private void deleteConfig(EntityId entityId) {
+    Transactionals.execute(transactional, context -> {
+      PreferencesDataset dataset = PreferencesDataset.get(context, datasetFramework);
+      Map<String, String> oldProp = dataset.getPreferences(entityId);
+      dataset.deleteProperties(entityId);
+      if (oldProp.containsKey(SystemArguments.PROFILE_NAME)) {
+        adminEventPublisher.publishProfileUnAssignment(messagingContext, entityId);
+      }
+    });
+  }
+
+  /**
+   * Get instance level preferences
+   */
+  public Map<String, String> getProperties() {
+    return getConfigProperties(new InstanceId(""));
+  }
+
+  /**
+   * Get namespace level preferences
+   */
+  public Map<String, String> getProperties(String namespace) {
+    return getConfigProperties(new NamespaceId(namespace));
+  }
+
+  /**
+   * Get app level preferences
+   */
+  public Map<String, String> getProperties(String namespace, String appId) {
+    return getConfigProperties(new ApplicationId(namespace, appId));
+  }
+
+  /**
+   * Get program level preferences
+   */
+  public Map<String, String> getProperties(String namespace, String appId, String programType, String programId) {
```

Comment:
it would be better for these methods to take NamespaceId, ApplicationId, ProgramId, as it would ensure that this class doesn't have to worry about invalid entities. 

Though if there are a lot of usages of these methods, it would be better to do that cleanup in a follow up PR.

Reply:
Fixed